### PR TITLE
Load balancing refactor

### PR DESCRIPTION
**Pull request by cvybhu**
This PR is a draft of how the new load balancing code could look like. This is a continuation of efforts by havaker in #449, although I deviated from his ideas.

## Design goals
* Get rid of policy composability - it made the code hard to understand and slow, while not being that useful
* Improve performance - the new design avoids any allocations on the fast path - the first few nodes in a plan are always outputted without allocating anything

## New traits
The interface has been changed a bit:
* `LoadBalancingPolicy` returns a special struct instead of `Box<dyn Iterator>`. This struct is an enum which can hold the common plan types implemented by the driver. This way we are able to avoid allocating the iterator when planning a query.
* A load balancing plan operates on `&Arc<Node>` instead of `Arc<Node>` to avoid needless atomic operations
* The plan iterator can borrow the `policy`, `statement_info` and `ClusterData`
```rust

pub struct StatementInfo<'a> {
    pub token: Option<Token>,
    pub keyspace: Option<&'a str>,
}

/// Policy that decides which nodes to contact for each query
pub trait LoadBalancingPolicy: Send + Sync {
    /// It is used for each query to find which nodes to query.\
    /// The generated plan can borrow any of the arguments for the whole query.
    fn plan<'a>(
        &'a self,
        statement_info: &'a StatementInfo,
        cluster: &'a ClusterData,
    ) -> LBPlan<'a>;

    /// Returns name of this load balancing policy
    fn name(&self) -> Cow<'_, str>;
}

/// Represents a load balancing plan generated by a load balancing policy.\
/// Common policies have their plan added as an enum to avoid needless allocations.\
/// In case of a custom policy it has to allocate a Box with its plan inside.
pub enum LBPlan<'a> {
    Dumb(DumbPlan<'a>),
    SimpleStrategyPlan(SimpleStrategyPlan<'a>),
    NoTokenNetworkStrategyPlan(NoTokenNetworkStrategyPlan<'a>),
    NetworkStrategyPlan(NetworkStrategyPlan<'a>),
    Custom(Box<dyn LoadBalancingPlan<'a> + Send + 'a>),
}

/// Each load balancing plan implements this trait.\
/// A plan is almost like an Iterator<Item = &Arc<Node>> but the lifetimes must be correct.\
pub trait LoadBalancingPlan<'a> {
    fn next(&mut self) -> Option<&'a Arc<Node>>;
}
```

## Precomputing replica sets
Currently replica sets are computed dynamically for each query plan. While walking the token ring we have too keep track of already visited nodes, which causes allocations. This refactor introduces precomputing of these sets. 

The driver looks at all existing keyspaces and finds maximum replication factors in the global ring and datacenter rings.
Then for each ring and each token it computes replica set for this ring's maximum replication factor.
Lower replication factor replica sets can be found by taking a prefix of maximum replication factor replicas.

This way we compute just enough data to serve any query instantly. When a replica set is needed the driver tries to find it in the precomputed data, but if for some reason the right replica set isn't currently available it falls back to dynamically generating one.

## `MainPolicy`
Instead of having multiple load balancing policies there's now only `MainPolicy`. It can be provided with a local datacenter name and it does the right thing™.
Do we even need other policies?

## Events?
The previous load balancing refactor PR (#449) considered adding events based on which load balancing policies can build their view of the cluster. Having events would also possibly remove the need to rebuild `ClusterData` from scratch each time cluster topology changes. 

I'm not really convinced that they should be introduced along with this load balancing refactor, they introduce a lot of complexity, while the benefits don't seem that big to me. Not having to rebuild `ClusterData` would be nice, but it doesn't take that long. Precomputed replicas would probably need to be recomputed anyway, or would require some complex and error prone algorithm for dynamic updates. Maybe @havaker has something to comment about them.

## TODO:

- [ ] Is this the right direction to go in?
- [ ] Is the code correct?
- [ ] Is it faster? How much?
- [ ] How do NetworkTopologyStrategy replicas work when a node doesn't have a rack?
- [ ] How long does precomputing replica sets take for a reasonably sized cluster? How much memory does it use?
- [ ] I do precomputing each time `ClusterData` is created, but that might be every 60 seconds when topology is updated (?) That would be really wasteful, it should only be recomputed when topology changes
- [ ] Do we need other policies than the Main one?
- [ ] Write documentation: comments, commit messages, book
- [ ] Write tests, maybe even integration ones?

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [ ] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [ ] PR description sums up the changes and reasons why they should be introduced.
- [ ] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by cvybhu** [comment]
There are merge conflicts because I started my branch before e66c6d8f20817a119d5f813eab191d286a110f1d to avoid problems with  #487, #490

**Comment by cvybhu** [comment]
cc @havaker @psarna @piodul

File: scylla/src/transport/load_balancing/dumb.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+use super::{LoadBalancingPolicy, Plan, Statement};
+use crate::transport::cluster::ClusterData;
+
+pub struct DumbPolicy;
```
Comment (psarna) [inline scylla/src/transport/load_balancing/dumb.rs:6]: We need an explanation for its existence (and a commit message), and strictly speaking this policy may not be "dumb" for certain workloads. We can call it "SimplePolicy" or "AllNodesPolicy".

File: scylla/src/transport/session.rs [hunk 2]
```diff
@@ -31,9 +31,7 @@ use crate::transport::connection::{
 };
 use crate::transport::connection_pool::PoolConfig;
 use crate::transport::iterator::{PreparedIteratorConfig, RowIterator};
-use crate::transport::load_balancing::{
-    LoadBalancingPolicy, RoundRobinPolicy, Statement, TokenAwarePolicy,
-};
+use crate::transport::load_balancing::{DumbPolicy, LoadBalancingPolicy, Statement};
```
Comment (psarna) [inline scylla/src/transport/session.rs]: Ah, ok, it's transitional. Still, each commit deserves a short commit message to introduce the reviewer into your line of thought

File: scylla/src/transport/iterator.rs [hunk 3]
```diff
@@ -29,7 +29,7 @@ use crate::statement::Consistency;
 use crate::statement::{prepared_statement::PreparedStatement, query::Query};
 use crate::transport::cluster::ClusterData;
 use crate::transport::connection::{Connection, QueryResponse};
-use crate::transport::load_balancing::{LoadBalancingPolicy, Statement};
+use crate::transport::load_balancing::{LoadBalancingPolicy, StatementInfo};
```
Comment (psarna) [inline scylla/src/transport/iterator.rs:33]: missing commit message with rationale why it's needed

File: scylla/src/transport/load_balancing/mod.rs [hunk 4]
```diff
@@ -19,13 +20,89 @@ pub struct StatementInfo<'a> {
...
+    fn from(plan: LBPlan<'a>) -> SharedLBPlan<'a> {
+        SharedLBPlan {
+            plan: Mutex::new(plan),
+        }
+    }
+}
+
+impl<'a> LoadBalancingPlan<'a> for &SharedLBPlan<'a> {
+    fn next(&mut self) -> Option<&'a Arc<Node>> {
+        let mut_ref = &mut self.plan.lock().unwrap();
```
Comment (psarna) [inline scylla/src/transport/load_balancing/mod.rs:109]: `unwrap()` in library code deserves an explanation why it's always true; and if it isn't, it should be some kind of a runtime error instead

**Review by psarna — COMMENTED** [review]
I stopped the review mid-series because lack of commit messages makes it considerably harder to follow what happens and why. It's fine for a draft though, so I'm just leaving a couple of comments for the commits I browsed.

**Comment by psarna** [comment]
> This way we compute just enough data to serve any query instantly. When a replica set is needed the driver tries to find it in the precomputed data, but if for some reason the right replica set isn't currently available it falls back to dynamically generating one.

Is the implementation fully lazy, i.e. it only recomputes once it finds out that some replica isn't available? That might make the pessimistic case considerably more costly, so perhaps we should also periodically recompute as well, to keep the information sufficiently up-to-date.

> Instead of having multiple load balancing policies there's now only MainPolicy. It can be provided with a local datacenter name and it does the right thing™.
> Do we even need other policies?

(I know I'm answering a rhetorical question here) Yup, we already had users that implemented their own policies for their specific use cases.


> Events?

I'm personally fine with dropping the idea of events if it simplifies the implementation.

**Comment by piodul** [comment]
@cvybhu can we close this? All of the work is being continued here: https://github.com/scylladb/scylla-rust-driver/pull/612 , right?

**Pull request closed without merging**